### PR TITLE
xds: delta-xDS avoid copying resources

### PR DESCRIPTION
**Pull request by adisuissa**
Commit Message: xds: avoid copying resources
Additional Description:
This PR avoids copying the non-heartbeat resources into their own vector, by partitioning the original repeated-field and passing the relevant range. This should reduce the amount of memory used when ingesting a new config.
No impact to the behavior is expected, but a runtime flag was added as this may impact config-ingestion related extensions (e.g., custom-config-validators, config-tracker), as the order of the elements passed to the callback functions may be different. This change can be temporarily reverted by setting `envoy.reloadable_features.xds_prevent_resource_copy` to `false`.
Risk Level: low
Testing: No behavior change, but added more tests.
Docs Changes: N/A
Release Notes: Added.
Platform Specific Features: N/A
Runtime guard: envoy.reloadable_features.xds_prevent_resource_copy

### CODE REVIEW

**Comment by adisuissa** [comment]
Assigning Stephan who also reviewed #36316
/assign @zuercher

File: source/extensions/config_subscription/grpc/xds_mux/delta_subscription_state.cc [hunk 1]
```diff
@@ -188,8 +190,24 @@ void DeltaSubscriptionState::handleGoodResponse(
...
 
-  absl::Span<const envoy::service::discovery::v3::Resource* const> non_heartbeat_resources_span =
-      absl::MakeConstSpan(non_heartbeat_resources.data(), non_heartbeat_resources.size());
+  absl::Span<const envoy::service::discovery::v3::Resource* const> non_heartbeat_resources_span;
+  if (Runtime::runtimeFeatureEnabled("envoy.reloadable_features.xds_prevent_resource_copy")) {
+    // Reorder the resources in the response, having all the non-heartbeat
+    // resources at the front of the list. Note that although there's no
+    // requirement to keep stable ordering, we do so to process the resources in
+    // the order they were sent.
+    auto last_non_heartbeat = std::stable_partition(
```
Comment (nezdolik) [inline source/extensions/config_subscription/grpc/xds_mux/delta_subscription_state.cc:199]: would two discovery responses with different order of resources but same resources considered equal?

**Review by nezdolik — APPROVED** [review]

File: source/extensions/config_subscription/grpc/xds_mux/delta_subscription_state.cc (hunk 1 above)
Comment (adisuissa) [reply source/extensions/config_subscription/grpc/xds_mux/delta_subscription_state.cc:199]: The protocol assumes that the given repeated field will be a set of resources, which is "enforced" by the having different resource names for the resources.
The implementation (in Envoy) "allows" having the "same resource" in the discovery-response, but will only take the first element in the list (e.g., [cds](https://github.com/envoyproxy/envoy/blob/312e01def6594155127d6d9a4e1a1de432551390/source/common/upstream/cds_api_helper.cc#L43-L48)).
In this PR I've attempted to keep the same order is done before the change by using a stable-partition.

**Review by zuercher — APPROVED** [review]

File: source/extensions/config_subscription/grpc/xds_mux/delta_subscription_state.cc (hunk 1 above)
Comment (nezdolik) [reply source/extensions/config_subscription/grpc/xds_mux/delta_subscription_state.cc:199]: thanks for explanation!

**Pull request merged**